src/messaging/ReliableMessageContext.cpp

```diff
@@ -0,0 +1,493 @@
+ *  the active retransmit timeout based on whether the ExchangeContext has
+ *  an active message exchange going with its peer.
+ *
+ *  @return the current retransmit time.
+ */
+uint64_t ReliableMessageContext::GetCurrentRetransmitTimeoutTick()
+{
+    return (HasRcvdMsgFromPeer() ? mConfig.mActiveRetransTimeoutTick : mConfig.mInitialRetransTimeoutTick);
+}
+
+/**
+ *  Send a Throttle Flow message to the peer node requesting it to throttle its sending of messages.
+ *
+ *  @note
+ *    This message is part of the CHIP Reliable Messaging protocol.
+ *
+ *  @param[in]    pauseTimeMillis    The time (in milliseconds) that the recipient is expected
+ *                                   to throttle its sending.
+ *  @retval  #CHIP_ERROR_INVALID_ARGUMENT               If an invalid argument was passed to this SendMessage API.
+ *  @retval  #CHIP_ERROR_SEND_THROTTLED                 If this exchange context has been throttled when using the
+ *                                                       CHIP reliable messaging protocol.
+ *  @retval  #CHIP_ERROR_WRONG_MSG_VERSION_FOR_EXCHANGE If there is a mismatch in the specific send operation and the
+ *                                                       CHIP message protocol version that is supported. For example,
+ *                                                       this error would be generated if CHIP Reliable Messaging
+ *                                                       semantics are being attempted when the CHIP message protocol
+ *                                                       version is V1.
+ *  @retval  #CHIP_ERROR_NOT_CONNECTED                  If the context was associated with a connection that is now
+ *                                                       closed.
+ *  @retval  #CHIP_ERROR_INCORRECT_STATE                If the state of the exchange context is incorrect.
+ *  @retval  #CHIP_NO_ERROR                             If the CHIP layer successfully sent the message down to the
+ *                                                       network layer.
+ *
+ */
+CHIP_ERROR ReliableMessageContext::SendThrottleFlow(uint32_t pauseTimeMillis)
+{
+    CHIP_ERROR err                = CHIP_NO_ERROR;
+    System::PacketBuffer * msgBuf = nullptr;
+    uint8_t * p                   = nullptr;
+    uint8_t msgLen                = sizeof(pauseTimeMillis);
+
+    msgBuf = System::PacketBuffer::NewWithAvailableSize(msgLen);
+    VerifyOrExit(msgBuf != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    p = msgBuf->Start();
+
+    // Encode the fields in the buffer
+    Encoding::LittleEndian::Write32(p, pauseTimeMillis);
+    msgBuf->SetDataLength(msgLen);
+
+    // Send a Throttle Flow message to the peer.  Throttle Flow messages must never request
+    // acknowledgment, so suppress the auto-request ACK feature on the exchange in case it has been
+    // enabled by the application.
+    err = ReliableMessageManager::GetManager().SendMessage(
```

Comment:
could we avoid singletons and instead make the message manager pointer a member of this class?

This comment applies throughout this file: ReliableMessageManager seems too coupled and used everywhere.

Reply:
done